### PR TITLE
staticMap: explain design

### PR DESCRIPTION
**Pull request by WalterBright**
It isn't obvious why the implementation strategy is used, so added an explanation.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
:+1:

**Comment by dnadlinger** [comment]
Hmm, wasn't the original design intent rather to reduce the total amount of mangled name bytes?

**Comment by adamdruppe** [comment]
I'd also probably argue there should be another threshold on that too. idk where, what N should be, but the simple recursion might actually be better if T.length < N, else try the divide and conquer.

Though the < 8 cases special-cased by hand might already handle this.

Or we could do a string mixin that works for all N reasonably well.

Better to comment the existing behavior than nothing though so I'd merge this regardless if i had the pwoer.

**Comment by dnadlinger** [comment]
> Hmm, wasn't the original design intent rather to reduce the total amount of mangled name bytes?

Misremembered that, this must have been some other template.

Is it square root, though, or log?

**Comment by WalterBright** [comment]
Maybe it's log2(n)

**Comment by FeepingCreature** [comment]
It's log_2. Each depth halves the set size, so given a depth n the total number of steps is 2^n.

File: std/meta.d [hunk 1]
```diff
@@ -857,6 +857,13 @@ template staticMap(alias F, T...)
 
     else
     {
+        /* While:
+         *   alias staticMap = aliasSeq!(F!T[0], staticMap!(F, T[0 .. $]));
```
Comment (schveiguy) [inline std/meta.d:861]: Even though it's an uncompiled comment, The second parameter should be `T[1 .. $]`

Comment (schveiguy) [reply std/meta.d:861]: And `AliasSeq` should be capitalized.

**Comment by adamdruppe** [comment]
> Hmm, wasn't the original design intent rather to reduce the total amount of mangled name bytes?

staticMap should never mangle to anything at all nor affect anything else, since it just yields an alias of an existing thing.

And as far as I can tell, it indeed doesn't in the generated binary. Not sure about inside dmd itself though but it probably shouldn't there either.

**Comment by schveiguy** [comment]
> Not sure about inside dmd itself though but it probably shouldn't there either.

It gets into the symbol table. This is why it consumes DMD memory. One of the benefits of possibly having the `...` improvement, or using type functions.

**Comment by dnadlinger** [comment]
> It's log_2. Each depth halves the set size, so given a depth n the total number of steps is 2^n.

That's what I said. Clearly I'm becoming too British for my own good. :P

**Comment by schveiguy** [comment]
This is trivial, I just made the changes myself.

**Pull request merged**